keras/src/backend/mlx/nn.py

```diff
@@ -122,16 +125,172 @@ def log_softmax(x, axis=-1):
     return x - mx.logsumexp(x, axis=axis, keepdims=True)
 
 
+def _calculate_padding(input_shape, pool_size, strides):
+    ndim = len(input_shape)
+
+    padding = ()
+    for d in range(ndim):
+        pad = max(0, (pool_size[d] - 1) - ((input_shape[d] - 1) % strides[d]))
+        padding = padding + (pad,)
+
+    return [(p // 2, (p + 1) // 2) for p in padding]
+
+
+def _non_overlapping_sliding_windows(x, shape, window_shape):
+    # Compute the intermediate shape
+    new_shape = [shape[0]]
+    for s, w in zip(shape[1:], window_shape):
+        new_shape.append(s // w)
+        new_shape.append(w)
+    new_shape.append(shape[-1])
+
+    last_axis = len(new_shape) - 1
+    axis_order = [
+        0,
+        *range(1, last_axis, 2),
+        *range(2, last_axis, 2),
+        last_axis,
+    ]
+
+    x = x.reshape(new_shape)
+    x = x.transpose(axis_order)
+    return x
+
+
+def _sliding_windows(x, window_shape, window_strides):
+    if x.ndim < 3:
+        raise ValueError(
+            f"To extract sliding windows at least 1 spatial dimension "
+            f"(3 total) is needed but the input only has {x.ndim} dimensions."
+        )
+
+    spatial_dims = x.shape[1:-1]
+    if not (len(spatial_dims) == len(window_shape) == len(window_strides)):
+        raise ValueError(
+            f"To extract sliding windows the window shapes and strides must "
+            f"have the same number of spatial dimensions as the signal but "
+            f"the signal has {len(spatial_dims)} dims and the window shape "
```

Comment:
Note: lines that don't have `{}`s in them don't need to start with `f`. This applies to all string lines in the PR.

In this error message, please format the argument information via `name={value}`, e.g.

"input_shape={x.shape}, window_shape={window_shape}, strides={window_strides}"